#' A function for processing abricate output.
#
#' Read the documents on the github page https://github.com/maxlcummins/abricateR for more info
#' @param file Filename of your abricate input
#' @param output Prefix for object/filename of your output files
#' @param identity Nucleotide identity cut-off. Default = 90
#' @param lenth Length/coverage cut-off. Default = 90
#' @param writecsv TRUE/FALSE: Write/Dont write to csv. Default = FALSE
#' @param pointfinder_data Defaults to a placeholding character - replace with a path to your concatenated pointfinder data.
#' @param ColV_Liu_data Defaults to a placeholding character - replace with a path to your abriate data for the Liu ColV database.
#' 
#' If you want to add pointfinder data, concatenate it as follows to provide filenames for the strains. This code assumes you are in a folder with only your ${file}.tab files generated by pointfinder.
#' for f in */*.tsv; do awk 'NR == 1 {print $0 "\tname_file"; next;}{print $0 "\t" FILENAME;}' $f; done > pointfinder_results_tmp.txt
#' awk 'FNR==1 { header = $0; print } $0 != header' pointfinder_results_tmp.txt > pointfinder_results.txt
#


abricateR <-
        function(file,
                 output,
                 identity = 90,
                 length = 95,
                 writecsv = FALSE,
                 pointfinder_data = "a",
                 ColV_Liu_data = "b") {
                
                require(readr)
                require(magrittr)
                require(dplyr)
                require(tidyverse)
                require(purrr)
                require(data.table)
                
                #provides a name for downstream files based on user input
                name <- output
                #reads input file (BLAST output in tab separated format.
                #                   Note: MUST BE GENERATED USING CUSTOM
                #                   PROVIDED BLAST SUBMISSION SCRIPT)
                #Colname reassignment
                
                
                #Read in full abricate genotype data sheet
                file <- read_delim(
                        file,
                        "\t",
                        escape_double = FALSE,
                        trim_ws = TRUE,
                        col_names = TRUE
                )
                
                #Remove cases where there are multiple headers from concatenation of abricate reports
                file <- file %>% filter(SEQUENCE != "SEQUENCE")
                
                #Colname reassignment
                colnames(file)[c(1, 10:11)] <-
                        c("name", "perc_coverage", "perc_identity")
                df_colnames <- colnames(file)
                
                #Convert percent coverage and identity to numeric type to allow filtering
                file$perc_coverage <- as.numeric(file$perc_coverage)
                file$perc_identity <- as.numeric(file$perc_identity)
                
                # appends a 1 to the final column of each row (used later by reshape2::dcast)
                rep(x = 1, times = nrow(file)) -> file$gene_present
                
                # Removes .fasta* in column sample name
                gsub(pattern = "\\.fasta.*",
                     replacement = "",
                     x = file$name) ->
                        file$name
                
                if (pointfinder_data != "a") {
                        #Read in pointfinder data on resistance snps
                        res_snps <- read_delim(
                                pointfinder_data,
                                "\t",
                                escape_double = FALSE,
                                trim_ws = TRUE
                        )
                        
                        #fix colnames of res_snps for consistency
                        colnames(res_snps) <-
                                gsub("name_file", "name", colnames(res_snps))
                        
                        #fix colnames of res_snps for consistency
                        colnames(res_snps) <-
                                gsub("Mutation", "GENE", colnames(res_snps))
                        
                        #fix colnames of res_snps for consistency
                        res_snps$GENE <-
                                gsub("^", "pointfinder_", res_snps$GENE)
                        
                        #fix sample names of res_snps for consistency
                        res_snps$name <- gsub("/.*.tsv", "", res_snps$name)
                        
                        # appends a 1 to the final column of each row (used later by reshape2::dcast)
                        rep(x = 1, times = nrow(res_snps)) -> res_snps$gene_present
                        
                        res_snps <- as.data.table(res_snps)
                        
                        #Cast the dataframe into a more usable format
                        res_snps2 <-
                                data.table::dcast(
                                        data = res_snps,
                                        formula = name ~ GENE,
                                        value.var = 'gene_present',
                                        drop = FALSE
                                )
                        
                }
                
                if (ColV_Liu_data != "b") {
                        #Read in abricate data on resistance snps
                        colv <- read_delim(ColV_Liu_data,
                                   "\t",
                                   escape_double = FALSE,
                                   trim_ws = TRUE)
                        
                        #Remove cases where there are multiple headers from concatenation of abricate reports
                        colv <- colv %>% filter(SEQUENCE != "SEQUENCE")
                        
                                                #Colname reassignment
                        colnames(colv)[c(1, 10:11)] <-
                                c("name", "perc_coverage", "perc_identity")
                        df_colnames <- colnames(colv)
                        
                        #Convert percent coverage and identity to numeric type to allow filtering
                        colv$perc_coverage <- as.numeric(colv$perc_coverage)
                        colv$perc_identity <- as.numeric(colv$perc_identity)
                        
                        #apply our filter to only allow genes which have a percent coverage of >=95 and a percent identity of >=90
                        colv <-
                                colv %>% filter(perc_coverage >= length) %>% filter(perc_identity >= identity)
                        
                        # appends a 1 to the final column of each row (used later by reshape2::dcast)
                        rep(x = 1, times = nrow(colv)) -> colv$gene_present
                        
                        # Removes .fasta* in column sample name
                        gsub(pattern = "\\.fasta.*",
                             replacement = "",
                             x = colv$name) ->
                                colv$name
                        
                        # appends a 1 to the final column of each row (used later by reshape2::dcast)
                        rep(x = 1, times = nrow(colv)) -> colv$gene_present
                        
                        #cast the data to generate a table showing which genes are present and which genes are absent
                        #i.e. make the list wide rather than long
                        simple_summary <- reshape2::dcast(
                                data = colv,
                                name ~ GENE,
                                value.var = 'gene_present',
                                drop = FALSE
                        )
                        
                        #create a new column for each group of genes and assign to each cell in this column the sum of the number of genes from each group present in a given sample
                        simple_summary <-
                                simple_summary %>% mutate(grp1_cvaABC_cvi = cvaA + cvaB + cvaC + cvi)
                        simple_summary <-
                                simple_summary %>% mutate(grp2_iroBCDEN = iroB + iroC + iroD + iroE + iroN)
                        simple_summary <-
                                simple_summary %>% mutate(grp3_iucABCD_iutA = iucA + iucB + iucC + iucD + iutA)
                        simple_summary <-
                                simple_summary %>% mutate(grp4_etsABC = etsA + etsB + etsC)
                        simple_summary <-
                                simple_summary %>% mutate(grp5_ompT_hlyF = ompT + hlyF)
                        simple_summary <-
                                simple_summary %>% mutate(grp6_sitABCD = sitA + sitB + sitC + sitD)
                        
                        #Save the row names to a different object so we can later reassign them after editing the table
                        assembly_name <- simple_summary$name
                        
                        #Take all rows and all columns except for the first column and reassign the dataframe "simple summary" as such
                        simple_summary <-
                                simple_summary[1:nrow(simple_summary), 2:ncol(simple_summary)]
                        
                        #Replace gene hits such that if any gene was detected more than once, or if more than one gene in a given group were detected, a corresponding gene/gene group is not counted twice
                        simple_summary[simple_summary > 1] <- 1
                        
                        #reassign the Assembly_barcode column
                        simple_summary$name <- assembly_name
                        
                        #change the column order so that Assembly_barcode comes first.
                        simple_summary <-
                                simple_summary %>% select(name, everything())
                        
                        #Add an extra column which sums the number of gene groups detected
                        simple_summary <-
                                simple_summary %>% mutate(
                                        grp_sum = grp1_cvaABC_cvi + grp2_iroBCDEN + grp3_iucABCD_iutA + grp4_etsABC +
                                                grp5_ompT_hlyF + grp6_sitABCD
                                )
                        
                        #If 3 or less gene groups were detected within a given sample, replace this value with No (ColV plasmid not considered present)
                        simple_summary$ColV <- gsub("1|2|3", "No", simple_summary$grp_sum)
                        
                        #If 4 or more gene groups were detected within a given sample, replace this value with Yes (ColV plasmid considered present)
                        simple_summary$ColV <- gsub("4|5|6", "Yes", simple_summary$ColV)
                        
                        simple_summary$ColV <- gsub("No", "0", simple_summary$ColV)
                        simple_summary$ColV <- gsub("Yes", "1", simple_summary$ColV)
                        
                        #Replace all NAs with zeroes, as these strains did not carry and genes from the ColV abricate screen at all and instead come up as NA
                        #CHECK THIS
                        simple_summary[is.na(simple_summary)] <- 0
                        
                        assign(paste(name, "colv", paste("N", identity, sep = ""), paste("L", length, sep =""), sep = "."), simple_summary, envir=globalenv())
                        assign(paste(name, "colv_summary", sep = "."), colv, envir=globalenv())
                        ColV_binary <- simple_summary %>% select(name, ColV)
                        
                }
                
                
                #adds the name of the database to the start of the gene name (for later sorting)
                file$GENE <- paste0(file$DATABASE, "_", file$GENE)
                
                # Assigns all hits meeting criteria of
                # X% Nucleotide match and Y% Length Match to variable filename.NX.LY.PASS
                # (tweak arguments 3 and 4 of blastlord to change the nucleotide and length criteria)
                file1 <-
                        subset.data.frame(file,
                                          file$perc_identity >= identity & file$perc_coverage >= length)
                assign(paste(
                        name,
                        paste("N", identity, sep = ""),
                        paste("L", length, sep = ""),
                        "PASS",
                        sep = "."
                ),
                file1,
                envir = globalenv())
                
                # Assigns all hits NOT meeting criteria of
                # X% Nucleotide match and Y% Length Match to variable filename.NX.LY.FAIL
                # (tweak arguments 3 and 4 of blastlord to change the nucleotide and length criteria)
                file2 <-
                        subset.data.frame(file,
                                          file$perc_identity < identity & file$perc_coverage < length)
                assign(paste(
                        name,
                        paste("N", identity, sep = ""),
                        paste("L", length, sep = ""),
                        "FAIL",
                        sep = "."
                ),
                file2,
                envir = globalenv())
                
                require("reshape2")
                
                
                
                # Generates a simple summary table of genes (not including allele variants) from
                # file1 (PASS file) that contains hits that passed the set criteria
                reshape2::dcast(
                        data = file1,
                        name ~ GENE,
                        value.var = 'gene_present',
                        drop = FALSE
                ) -> file3
                if (pointfinder_data != "a") {
                        file3 <- left_join(file3, res_snps2)
                        file3[is.na(file3)] <- 0 
                        message("Combining pointfinder data...")
                }
                if (ColV_Liu_data != "b") {
                        file3 <- left_join(file3, ColV_binary)
                        file3[is.na(file3)] <- 0 
                        message("Combining pointfinder data...")
                }
                assign(
                        paste(
                                name,
                                "_simple_summary_",
                                "N",
                                identity,
                                "L",
                                length,
                                sep = ""
                        ),
                        file3,
                        envir = globalenv()
                )
                
                # Generates a table showing co-occurence of all genes on a given SEQUENCE for each sample.
                file5 <- file1 %>%
                        group_by(name, SEQUENCE) %>%
                        summarise(same_scaff = paste(unique(GENE), collapse = " "))
                assign(
                        paste(name, "_co-occurence_", "N", identity, "L", length, sep = ""),
                        file5,
                        envir = globalenv()
                )
                
                if (writecsv == TRUE) {
                        message("Writing objects to csv file...")
                        write.csv(file1, paste(
                                paste(
                                        name,
                                        paste("N", identity, sep = ""),
                                        paste("L", length, sep = ""),
                                        "PASS",
                                        sep = "."
                                ),
                                ".csv",
                                sep = ""
                        ))
                        write.csv(file2, paste(
                                paste(
                                        name,
                                        paste("N", identity, sep = ""),
                                        paste("L", length, sep = ""),
                                        "FAIL",
                                        sep = "."
                                ),
                                ".csv",
                                sep = ""
                        ))
                        write.csv(file3, paste(
                                paste(
                                        name,
                                        "_simple_summary_",
                                        "N",
                                        identity,
                                        "L",
                                        length,
                                        sep = ""
                                ),
                                ".csv",
                                sep = ""
                        ))
                        write.csv(file5, paste(
                                paste(
                                        name,
                                        "_same_scaff_",
                                        "N",
                                        identity,
                                        "L",
                                        length,
                                        sep = ""
                                ),
                                ".csv",
                                sep = ""
                        ))
                        message("Writing complete; script finished.")
                }
                else {
                        message("Files not written to disk; script finished.")
                }
        }
